### PR TITLE
Complete refactoring of the API

### PR DESCRIPTION
**Pull request by sylvainar**
This pull request was created to discuss about the new version of the API. 

The code is totally refactored in order to make smart calls to external APIs (Wikidata, Wikipedia, Wikivoyage), making it really faster, without depending of the number of POIs loaded.

Feel free to make any comment before it's merged :wink: 

TODO : 
- [x] Making the code easier to read by separating functions and put them in different files
- [x] Reorder API requests to make it more efficient
- [x] Move params into a config file
- [x] Test it, check for breaking changes
- [x] Update doc

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 require 'multiCurl.php';
 
-error_reporting(0); // No need error reporting, or else it will crash the JSON export
+error_reporting(E_ALL); // No need error reporting, or else it will crash the JSON export
```
Comment (tikiki) [inline index.php]: This should be configurable in a separate file, to allow debugging easily.

File: index.php [hunk 2]
```diff
@@ -8,8 +8,10 @@
...
...fr','frp','frr','fur','fy','ga','gag','gan','gd','gl','glk','gn','gom','got','gu','gv','ha','hak','haw','he','hi','hif','ho','hr','hsb','ht','hu','hy','hz','ia','id','ie','ig','ii','ik','ilo','io','is','it','iu','ja','jbo','jv','ka','kaa','kab','kbd','kg','ki','kj','kk','kl','km','kn','ko','koi','kr','krc','ks','ksh','ku','kv','kw','ky','la','lad','lb','lbe','lez','lg','li','lij','lmo','ln','lo','lrc','lt','ltg','lv','mai','map-bms','mdf','mg','mh','mhr','mi','min','mk','ml','mn','mo','mr','mrj','ms','mt','mus','mwl','my','myv','mzn','na','nah','nap','nds','nds-nl','ne','new','ng','nl','nn','no','nov','nrm','nso','nv','ny','oc','om','or','os','pa','pag','pam','pap','pcd','pdc','pfl','pi','pih','pl','pms','pnb','pnt','ps','pt','qu','rm','rmy','rn','ro','roa-rup','roa-tara','ru','rue','rw','sa','sah','sc','scn','sco','sd','se','sg','sh','si','simple','sk','sl','sm','sn','so','sq','sr','srn','ss','st','stq','su','sv','sw','szl','ta','te','tet','tg','th','ti','tk','tl','tn','to','tpi','tr','ts','tt','tum','tw','ty','tyv','udm','ug','uk','ur','uz','ve','vec','vep','vi','vls','vo','wa','war','wo','wuu','xal','xh','xmf','yi','yo','za','zea','zh','zh-classical','zh-min-nan','zh-yue','zu');
```
Comment (nymous) [inline index.php]: Can we move this huuuuuuge array to a separate file? (just wondering)

File: index.php [hunk 3]
```diff
@@ -143,217 +144,33 @@
...
-					}
-				}
-			}
-		}
+	// Call the magic function, check for error, and push in output array
+	$temp_POI_output = getAndParseWikipediaPOI($language, $user_latitude, $user_longitude, $range, $maxPOI, $thumbnailWidth);
+	if(!array_key_exists('error',$temp_POI_output))
+	{
+		$output['poi']['poi_info'] = $temp_POI_output;
+		$output['poi']['nb_poi'] = count($output['poi']['poi_info']);
```
Comment (nymous) [inline index.php]: Invert these two lines to look the same as l.156-157

File: index.php [hunk 4]
```diff
@@ -8,8 +8,10 @@
 */
 
 require 'multiCurl.php';
+require 'getAndParseWikipediaPOI.php';
+require 'getAndParseWikivoyageGuides.php';
 
-error_reporting(0); // No need error reporting, or else it will crash the JSON export
+error_reporting(E_ALL); // No need error reporting, or else it will crash the JSON export
```
Comment (nymous) [inline index.php]: We should move this to a config file, to allow easier debugging (and other conf we may add in the future).

Comment (sylvainar) [inline index.php]: Nice idea, it will comes with the refactoring.

File: getAndParseWikivoyageGuides.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+function getAndParseWikivoyageGuides($language, $user_latitude, $user_longitude)
+{
+	// ===> Make the URL 
+	$wikivoyageRequest = 'https://'.$language.'.wikivoyage.org/w/api.php?action=query&format=json&' // Base
+	.'prop=coordinates|info|pageterms|pageimages&' // Props list
+	.'piprop=thumbnail&pithumbsize=144&pilimit=50&inprop=url&wbptterms=description' // Properties dedicated to image, url and description
+	."&generator=geosearch&ggscoord=$user_latitude|$user_longitude&ggsradius=10000&ggslimit=50"; // Properties dedicated to geosearch
```
Comment (nymous) [inline getAndParseWikivoyageGuides.php]: Why is the ggsradius (I assume this is the search range) hard coded?

File: getAndParseWikipediaPOI.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		$output_array[$currentPOIdata['pageid']]['wikipedia_id'] = $currentPOIdata['pageid'];
+	}
+	$wikipedia_pagesid_list = substr($wikipedia_pagesid_list, 1);
+
+	// ===> Now we got a list of wikipedia pages so we call Wikipedia API again to get infos on those pages
+	$apidata_request_wikipedia_info = "https://".$language.".wikipedia.org/w/api.php?format=json&action=query&prop=pageprops|info|pageimages&inprop=url&pilimit=1000&pithumbsize=".$thumbnailWidth."&pageids=".$wikipedia_pagesid_list;
+	if(!($apidata_json_wikipedia_info = @file_get_contents($apidata_request_wikipedia_info)))
+		return array("error" => "Wikipedia API is unreachable");
+
+	$apidata_array_wikipedia_info = json_decode($apidata_json_wikipedia_info,1);
```
Comment (nymous) [inline getAndParseWikipediaPOI.php]: Why is it sometimes json_decode(json, true) and sometimes json_decode(json, 1) ?

Comment (sylvainar) [inline getAndParseWikipediaPOI.php]: It's the same, but right, i'm going to uniform.

**Comment by sylvainar** [comment]
It's over. It's done.

**Pull request merged**